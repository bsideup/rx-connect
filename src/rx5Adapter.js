import { Observable, Subject, BehaviorSubject, ReplaySubject } from "rxjs-v5";
import "rxjs-v5/add/observable/of";
import "rxjs-v5/add/observable/interval";
import "rxjs-v5/add/observable/merge";
import { scan, debounce, map, pluck, take } from "rxjs-v5/operators";
import ObservableSymbol from "symbol-observable";

const handler = {
  get: function(obj, key) {
    return obj.hasOwnProperty(key) ? obj[key] : Observable;
  }
};

const ObservableProxy = new Proxy({
  scan,
  debounce,
  map,
  pluck,
  take,
  interval: Observable.interval,
  merge: Observable.merge,
  of: Observable.of
}, handler);

export default {

    Rx: {
      Observable: ObservableProxy,
      Subject,
      ReplaySubject,
      BehaviorSubject,
    },

    next(o, value) {
        o.next(value);
    },

    isObservable(obj) {
        return obj && typeof obj[ObservableSymbol] === 'function';
    },

    unsubscribe(subscription) {
        subscription.unsubscribe();
    }
}
