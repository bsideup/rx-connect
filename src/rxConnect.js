import React from "react";
import isPlainObject from "lodash.isplainobject";
import isObject from "lodash.isobject";
import isArray from "lodash.isarray";

const DEFAULT_OPTIONS = {
    noDebounce: false
};

function getDisplayName(WrappedComponent) {
    return WrappedComponent.displayName || WrappedComponent.name || "Component";
}
function isObservable(obj) {
    if (!obj) {
        return false;
    }

    return getAdapter().isObservable(obj);
}

export function getAdapter() {
    const adapter = rxConnect.adapter || require("./rx6adapter");
    return adapter.__esModule ? adapter.default : adapter;
}

export default function rxConnect(selector, options = DEFAULT_OPTIONS) {
    return WrappedComponent => class RxConnector extends React.PureComponent {

        static displayName = "RxConnector";

        stateSubscription = undefined;

        shouldDebounce = false;

        subProps = {};

        constructor(props) {
            super(props);
            const { Rx } = getAdapter();

            this.props$ = new (Rx.BehaviorSubject)(props);
        }

        componentWillMount() {
            const { Rx: { Observable } } = getAdapter();

            this.shouldDebounce = false;

            let mutations$ = selector;
            if(!isObservable(mutations$)) {
                if (typeof selector === 'function') {
                    mutations$ = selector(this.props$);
                } else {
                    // eslint-disable-next-line no-console
                    console.error(`Selector must be a function or an Observable. Check rxConnect of ${getDisplayName(WrappedComponent)}. Got: `, selector);
                    return;
                }
            }

            if(!isObservable(mutations$)) {
                    // eslint-disable-next-line no-undef
                if (mutations$ && typeof mutations$[Symbol.iterator] === 'function') {
                    mutations$ = Observable.merge(...mutations$);
                } else {
                    // eslint-disable-next-line no-console
                    console.error(`Selector must return an Observable or an iterator of observables. Check rxConnect of ${getDisplayName(WrappedComponent)}. Got: `, mutations$);
                    return;
                }
            }
            
            this.stateSubscription = mutations$.pipe(
                Observable.scan((state, mutation) => {
                    if (typeof mutation === "function") {
                        return mutation(state);
                    }

                    if (isPlainObject(mutation)) {
                        return Object.assign({}, state, mutation);
                    }

                    if (isObject(mutation) && !isArray(mutation)) {
                        return Object.assign({}, state, {...mutation});
                    }

                    // eslint-disable-next-line no-console
                    console.error(`Mutation must be a plain object or function. Check rxConnect of ${getDisplayName(WrappedComponent)}. Got: `, mutation);
                    return state;
                }, {}),
                Observable.debounce(() => (!options.noDebounce && this.shouldDebounce) ? Observable.interval(0) : Observable.of()),
                ).subscribe(state => {
                  this.subProps = state;
                  this.forceUpdate();
                });
        }

        componentDidMount() {
            this.shouldDebounce = true;
        }

        componentWillUnmount() {
            getAdapter().unsubscribe(this.stateSubscription);
        }

        componentWillReceiveProps(nextProps) {
            getAdapter().next(this.props$, nextProps);
        }

        render() {
            return React.createElement(WrappedComponent, this.subProps, this.props.children);
        }
    };
}
