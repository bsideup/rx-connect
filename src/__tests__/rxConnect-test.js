import React from "react";
import renderer from "react-test-renderer";

import { rxConnect } from "../";
import { getAdapter } from "../rxConnect";
import rx5Adapter from "../rx5Adapter";

const suites = {
    "RxJS 6": () => {},
    "RxJS 5": () => rxConnect.adapter = rx5Adapter,
}

Object.entries(suites).forEach(([ name, initializer ]) => describe(name, () => {
    let Rx;
    beforeEach(() => {
        initializer();
        const adapter = getAdapter();
        Rx = adapter.Rx;
    });

    it("works with Observable", () => {
        const props$ = Rx.Observable.of({ a: 123 });

        const Component = rxConnect(props$)(({ a }) => <div>{a}</div>);

        const tree = renderer.create(<Component />).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it("works with Array", () => {
        const props$ = () => [
            Rx.Observable.of({ a: 123 }),
            Rx.Observable.of({ foo: "bar" }),
        ];

        const Component = rxConnect(props$)(({ a, foo }) => <div>{a}{foo}</div>);

        const tree = renderer.create(<Component />).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it("works with Generator", () => {
        const props$ = function* () {
            yield Rx.Observable.of({ a: 123 });
            yield Rx.Observable.of({ foo: "bar" });
        };

        const Component = rxConnect(props$)(({ a, foo }) => <div>{a}{foo}</div>);

        const tree = renderer.create(<Component />).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it("passes properties as Observable", () => {
        const connector = props$ => props$.pipe(
          Rx.Observable.pluck("someProp"), 
          Rx.Observable.map(a => ({ a }))
        );

        const Component = rxConnect(connector)(({ a }) => <div>{a}</div>);

        const tree = renderer.create(<Component someProp={123} />).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it("passes children automatically", () => {
        const Component = rxConnect(Rx.Observable.of({}))(({ children }) => <div>{children}</div>);

        const tree = renderer.create(<Component>Hello, RxConnect!</Component>).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it("ignores not connect properties", () => {
        const Component = rxConnect(Rx.Observable.of({ }))(({ a }) => <div>{a}</div>);

        const tree = renderer.create(<Component a={123} />).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it("accepts function-based mutations", () => {
        const Component = rxConnect(Rx.Observable.of(() => ({ a: 123 })))(({ a }) => <div>{a}</div>);

        const tree = renderer.create(<Component />).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it("throws an error if mutation is neither an object or function", () => {
        // eslint-disable-next-line no-console
        console.error = jest.genMockFn();

        const Component = rxConnect(Rx.Observable.of([ 123 ]))(({ a }) => <div>{a}</div>);
        renderer.create(<Component />).toJSON();

        // eslint-disable-next-line no-console
        expect(console.error.mock.calls[0]).toMatchSnapshot();
    });

    it("throws an error if selector is neither an Observable or function", () => {
        // eslint-disable-next-line no-console
        console.error = jest.genMockFn();

        const Component = rxConnect(undefined)(({ a }) => <div>{a}</div>);
        renderer.create(<Component />).toJSON();

        // eslint-disable-next-line no-console
        expect(console.error.mock.calls[0]).toMatchSnapshot();
    });

    it("throws an error if selector returns non-Observable", () => {
        // eslint-disable-next-line no-console
        console.error = jest.genMockFn();

        const Component = rxConnect(() => undefined)(({ a }) => <div>{a}</div>);
        renderer.create(<Component />).toJSON();

        // eslint-disable-next-line no-console
        expect(console.error.mock.calls[0]).toMatchSnapshot();
    });

    it("receives new props", async () => {
        const connector = props$ => props$.pipe(
          Rx.Observable.map(({ a, b }) => ({ a: a + b }))
        );

        const Component = rxConnect(connector)(({ a }) => <div>{a}</div>);

        class Parent extends React.Component {
            state = {
                a: 10,
                b: 5
            };

            render() {
                return <Component {...this.state} />
            }
        }

        const parent = renderer.create(<Parent />);

        // 15
        expect(parent.toJSON()).toMatchSnapshot();

        parent.getInstance().setState({ a: -5 });

        // Still 15 because of debouncing
        expect(parent.toJSON()).toMatchSnapshot();

        parent.getInstance().setState({ b: -10 });

        // Skip 1 frame because of debounce
        await Rx.Observable.interval(0).pipe(
          Rx.Observable.take(1)
        ).toPromise();

        expect(parent.toJSON()).toMatchSnapshot();
    });

    it("noDebounce", () => {
        const Component = rxConnect(props$ => props$, { noDebounce: true })(({ i }) => <div>{i}</div>);

        class Parent extends React.Component {
            state = {
                i: 0
            };

            render() {
                return <Component {...this.state} />
            }
        }

        const parent = renderer.create(<Parent />);

        expect(parent.toJSON()).toMatchSnapshot();

        parent.getInstance().setState({ i: 1 });

        expect(parent.toJSON()).toMatchSnapshot();
    })

    it("handles unmount", () => {
        const props$ = new Rx.ReplaySubject();

        expect(props$.observers).toHaveLength(0);

        const Component = rxConnect(props$)(({ a }) => <div>{a}</div>);

        const node = renderer.create(<Component />);

        expect(props$.observers).toHaveLength(1);

        node.unmount();

        expect(props$.observers).toHaveLength(0);
    });
}));
