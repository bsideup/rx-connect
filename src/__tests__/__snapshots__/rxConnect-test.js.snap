exports[`RxJS 5 accepts function-based mutations 1`] = `
<div>
  123
</div>
`;

exports[`RxJS 5 ignores not connect properties 1`] = `<div />`;

exports[`RxJS 5 noDebounce 1`] = `
<div>
  0
</div>
`;

exports[`RxJS 5 noDebounce 2`] = `
<div>
  1
</div>
`;

exports[`RxJS 5 passes children automatically 1`] = `
<div>
  Hello, RxConnect!
</div>
`;

exports[`RxJS 5 passes properties as Observable 1`] = `
<div>
  123
</div>
`;

exports[`RxJS 5 receives new props 1`] = `
<div>
  15
</div>
`;

exports[`RxJS 5 receives new props 2`] = `
<div>
  15
</div>
`;

exports[`RxJS 5 receives new props 3`] = `
<div>
  -15
</div>
`;

exports[`RxJS 5 throws an error if mutation is neither an object or function 1`] = `
Array [
  "Mutation must be a plain object or function. Check rxConnect of Component. Got: ",
  Array [
    123,
  ],
]
`;

exports[`RxJS 5 throws an error if selector is neither an Observable or function 1`] = `
Array [
  "Selector must be a function or an Observable. Check rxConnect of Component. Got: ",
  undefined,
]
`;

exports[`RxJS 5 throws an error if selector returns non-Observable 1`] = `
Array [
  "Selector must return an Observable or an iterator of observables. Check rxConnect of Component. Got: ",
  undefined,
]
`;

exports[`RxJS 5 works with Array 1`] = `
<div>
  123
  bar
</div>
`;

exports[`RxJS 5 works with Generator 1`] = `
<div>
  123
  bar
</div>
`;

exports[`RxJS 5 works with Observable 1`] = `
<div>
  123
</div>
`;

exports[`RxJS 6 accepts function-based mutations 1`] = `
<div>
  123
</div>
`;

exports[`RxJS 6 ignores not connect properties 1`] = `<div />`;

exports[`RxJS 6 noDebounce 1`] = `
<div>
  0
</div>
`;

exports[`RxJS 6 noDebounce 2`] = `
<div>
  1
</div>
`;

exports[`RxJS 6 passes children automatically 1`] = `
<div>
  Hello, RxConnect!
</div>
`;

exports[`RxJS 6 passes properties as Observable 1`] = `
<div>
  123
</div>
`;

exports[`RxJS 6 receives new props 1`] = `
<div>
  15
</div>
`;

exports[`RxJS 6 receives new props 2`] = `
<div>
  15
</div>
`;

exports[`RxJS 6 receives new props 3`] = `
<div>
  -15
</div>
`;

exports[`RxJS 6 throws an error if mutation is neither an object or function 1`] = `
Array [
  "Mutation must be a plain object or function. Check rxConnect of Component. Got: ",
  Array [
    123,
  ],
]
`;

exports[`RxJS 6 throws an error if selector is neither an Observable or function 1`] = `
Array [
  "Selector must be a function or an Observable. Check rxConnect of Component. Got: ",
  undefined,
]
`;

exports[`RxJS 6 throws an error if selector returns non-Observable 1`] = `
Array [
  "Selector must return an Observable or an iterator of observables. Check rxConnect of Component. Got: ",
  undefined,
]
`;

exports[`RxJS 6 works with Array 1`] = `
<div>
  123
  bar
</div>
`;

exports[`RxJS 6 works with Generator 1`] = `
<div>
  123
  bar
</div>
`;

exports[`RxJS 6 works with Observable 1`] = `
<div>
  123
</div>
`;
